### PR TITLE
feat: MCP Integration (relates #79)

### PR DESCRIPTION
**Pull request by ECHO-HELLO-WORLD424**
## Summary 

This PR (relates to #79) adds initial MCP integration to NeuroPilot, enabling connection to MCP servers and using their tools as Neuro actions.

## Current Capabilities

- Connect Neuro to a single MCP server,,
- Basic permission control (autopilot, copilot and off, enable/disable all MCP tools),

## Current Limitations

- No multi MCP server support
- No unit test suite yet (that would require implementing mock mcp server, not implemented yet.)
- Only support Streamable-HTTP via `/mcp` now, no support for `stdio` or backward support for `/sse`
- No automatic MCP server discovery (You need to specify URL),
- Can't do Individual tool level dynamic register/unregister (You can only register/un-register all MCP tool now. This is bad and should be improved because although you can control tool use with manual approval but  too many tools will likely do harm to Neuro's context  and cause LLM performance drop)

### How It Works

The MCP integration creates a translation layer between the Model Context Protocol and NeuroPilot's custom Neuro API. Here's the flow:

```
MCP Server -> MCPManager -> RCEAction wrappers -> Neuro API -> Neuro AI
                                                                 |
                                  VSCode <- Tool Execution <- RCEActionHandler
```

1. **Connection**: Extension connects to an MCP server (`/mcp` endpoint)
2. **Tool Discovery**: MCPClient queries available tools via `tools/list`
3. **Translation**: Each MCP tool is translated to a Neuro action with:
   - Sanitized name (e.g., `mcp__fetch_url` -> `mcp_fetch_url`)
   - JSON schema from MCP tool's `inputSchema`
   - Permission requirement: `mcpTools`
   - Default permission level: `Copilot` 
4. **Registration**: Actions are registered with the Neuro client
5. **Execution**: When Neuro calls an MCP tool:
   - RCEActionHandler identifies it as an MCP action
   - Validates permissions and schema
   - Shows confirmation dialog (Copilot mode) or executes directly (Autopilot mode)
   - MCPManager forwards the call to the MCP server via `tools/call`
   - Result is sent back to Neuro

#### New Files (src/mcp/)

- **`mcp_manager.ts`**: Manages single MCP server connection, tool discovery, and execution routing
- **`mcp_actions.ts`**: Handles registration/unregistration of MCP tools as Neuro actions, manages `mcpActions` registry
- **`mcp_commands.ts`**: VSCode commands for user interaction (connect, disconnect, refresh, status, test)

- **`mcp_client.ts`**: Low-level HTTP client for MCP protocol communication
- **`translation.ts`**: Converts MCP tool schemas to Neuro action format (modified to handle edge cases)
- **`registry.ts`**: Maps sanitized Neuro action names to original MCP tool names (enhanced)
- **`index.ts`**: Module exports (updated)

#### Integration Points

- **`src/rce.ts`**: Modified `RCEActionHandler` to check for and execute MCP actions
  - Added `isMCPAction()` check in action lookup
  - Added branch to retrieve MCP actions from `mcpActions` registry
- **`src/config.ts`**: Added MCP configuration class with settings:
  - `mcp.enabled`: Enable/disable MCP integration
  - `mcp.serverUrl`: MCP server endpoint
  - `mcp.timeout`: Connection timeout
  - `mcp.autoConnect`: Auto-connect on Neuro client connection
  - New permission: `mcpTools` for MCP tool execution
- **`src/desktop/extension.ts`**: Extension activation registers MCP commands and auto-connect
- **`package.json`**: Added 5 new commands and 4 configuration settings

### Configuration

Users can configure MCP in VSCode settings:

```json
{
  "neuropilot.mcp.enabled": false,
  "neuropilot.mcp.serverUrl": "http://localhost:3000",
  "neuropilot.mcp.timeout": 30000,
  "neuropilot.mcp.autoConnect": true
}
```

Permission level for MCP tools is controlled via:

```json
{
  "neuropilot.permission.mcpTools": "copilot" // or "autopilot" or "off"
}
```

### User Commands

Available via Command Palette (Ctrl+Shift+P):

- **NeuroPilot MCP: Connect to MCP Server** - Connect to configured server
- **NeuroPilot MCP: Disconnect from MCP Server** - Disconnect and unregister tools
- **NeuroPilot MCP: Refresh MCP Tools** - Re-fetch tools from server
- **NeuroPilot MCP: Show MCP Status** - Display connection info and registered tools
- **NeuroPilot MCP: Test MCP Tool Call (Debug)** - Manually invoke a tool for debugging

### Quick Test

To test the integration:

1. Start an MCP server
2. Configure MCP server URL in extention settings
3. `Ctrl+Shift+P`, then find&execute "NeuroPilot MCP: Connect to MCP Server" command
4. Use "NeuroPilot MCP: Test MCP Tool Call (Debug)" command to verify direct communication between extension and MCP server
5. Try invoke mcp tools from test Neuro server

### CODE REVIEW

**Comment by KTrain5169** [comment]
converted to draft bc well thats what it is

File: package.json [hunk 1]
```diff
@@ -171,6 +171,31 @@
...
+			},
+			{
+				"command": "neuropilot.mcp.connectServer",
+				"title": "Connect to MCP Server",
+				"category": "NeuroPilot MCP"
+			},
+			{
+				"command": "neuropilot.mcp.disconnectServer",
+				"title": "Disconnect from MCP Server",
+				"category": "NeuroPilot MCP"
```
Comment (KTrain5169) [inline package.json]:
```suggestion
				"category": "NeuroPilot"
```

File: package.json [hunk 2]
```diff
@@ -171,6 +171,31 @@
...
+			},
+			{
+				"command": "neuropilot.mcp.refreshTools",
+				"title": "Refresh MCP Tools",
+				"category": "NeuroPilot MCP"
```
Comment (KTrain5169) [inline package.json]:
```suggestion
				"category": "NeuroPilot"
```

File: package.json [hunk 3]
```diff
@@ -914,6 +954,28 @@
 						"type": "boolean",
 						"default": true,
 						"markdownDescription": "Whether or not to toggle schemas that may use \"unsupported\" schema properties.\n\nRequires reloading permissions if changed."
+					},
+					"neuropilot.mcp.enabled": {
```
Comment (KTrain5169) [inline package.json]: We already have `neuropilot.permission.mcpTools`, use that as a check instead.

File: src/mcp/translation.ts [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+import type { JSONSchema7 } from 'json-schema';
+import type { Tool } from '@modelcontextprotocol/sdk/types.js';
+
+/**
+ * JSON Schema keywords that are forbidden in Neuro action schemas.
+ * See https://github.com/VedalAI/neuro-game-sdk/blob/main/API/SPECIFICATION.md#action
+ * These should be removed to ensure compatibility with Neuro.
+ */
+export const FORBIDDEN_SCHEMA_KEYS = new Set([
```
Comment (KTrain5169) [inline src/mcp/translation.ts]: Vedal has said at least some of these *can* be used but shouldn't be relied on.

File: package.json [hunk 5]
```diff
@@ -914,6 +954,28 @@
...
+						"markdownDescription": "The URL of the MCP server to connect to (e.g., `http://localhost:3000`).\n\nUse the **Connect to MCP Server** command to connect after setting this.\n\nFor multiple MCP servers, use a relay server that aggregates them."
+					},
+					"neuropilot.mcp.timeout": {
+						"type": "number",
+						"default": 30000,
+						"minimum": 1000,
+						"maximum": 300000,
+						"markdownDescription": "Timeout in milliseconds for MCP tool execution (1000-300000ms).\n\nDefault: 30000 (30 seconds)"
+					},
+					"neuropilot.mcp.autoConnect": {
```
Comment (KTrain5169) [inline package.json]: I don't know if this is a good idea, servers could abuse this to attempt prompt injection attacks against Neuro. Also, it wastes resources if left unused.

Then again, it's togglable, so I guess that's fine probably.

**Review by KTrain5169 — CHANGES_REQUESTED** [review]
preliminary review, haven't been able to actually test the changes yet

**Comment by ECHO-HELLO-WORLD424** [comment]
## Changes (Nov. 8)

Addressed some issues in the review. Let me know if other adjustments are needed!

### Fixed:

- Remove "Neuro MCP" things
  
- Use `neuropilot.permission.mcpTools`instead of `neuropilot.mcp.enabled` for all MCP features
  
- Remove `"@cfworker/json-schema`, use RCE to check schema
  
- Add changes to `CHANGELOG.md`
  

### Concerns about prompt injection:

- Now all MCP features are default to "Off", so it can't auto connect to MCP server without user approval.
  
- With that been said, we still must be careful and only use trusted MCP servers.
  

### About added web extension

- In theory it will work, but lots of MCP servers' CORS support is bad, which means that browser will likely block the connection attempt and the integration will produce a "can't fetch" error.
  
- So it works in theory but is less practical in real world for now.
  

### Notify Neuro about MCP tools

- A `mcp_` prefix will be added to mcp tool name.
  
- The extension will send notifications to Neuro on MCP connected/disconnected:
  
  - On MCP connected: "An MCP server with xx tools is now connected. You may access these tools whose names start with `mcp_`."
    
  - On MCP disconnected: "The MCP server is now disconnected. You can't access the tools whose names start with `mcp_` anymore."

File: src/web/extension.ts [hunk 6]
```diff
@@ -36,11 +37,14 @@ export function activate(context: vscode.ExtensionContext) {
...
 
+    // Register MCP commands
+    NEURO.context!.subscriptions.push(...registerMCPCommands(context));
+
     // Setup event handlers
     NEURO.context!.subscriptions.push(...setupCommonEventHandlers());
 
     // Setup client connected handlers
-    setupClientConnectedHandlers(registerUnsupervisedActions, registerUnsupervisedHandlers);
+    setupClientConnectedHandlers(registerUnsupervisedActions, registerUnsupervisedHandlers, autoConnectMCP);
```
Comment (KTrain5169) [inline src/web/extension.ts]: You also have this in `src/desktop/extension.ts`, so you should instead move them to `src/shared/extension.ts`.

File: src/web/extension.ts [hunk 7]
```diff
@@ -36,11 +37,14 @@ export function activate(context: vscode.ExtensionContext) {
     // Register commands
     NEURO.context!.subscriptions.push(...registerCommonCommands());
 
+    // Register MCP commands
+    NEURO.context!.subscriptions.push(...registerMCPCommands(context));
```
Comment (KTrain5169) [inline src/web/extension.ts]: Same for this

File: src/desktop/extension.ts [hunk 8]
```diff
@@ -51,7 +55,7 @@ export function activate(context: vscode.ExtensionContext) {
     registerChatParticipant();
 
     // Setup client connected handlers
-    setupClientConnectedHandlers(reloadTasks, registerUnsupervisedActions, registerUnsupervisedHandlers); // reloadTasks added to set it up at the same time
+    setupClientConnectedHandlers(reloadTasks, registerUnsupervisedActions, registerUnsupervisedHandlers, autoConnectMCP); // reloadTasks added to set it up at the same time
```
[the remark this answers is not in the document]
Comment (ECHO-HELLO-WORLD424) [reply src/desktop/extension.ts]: Possible but with limitations. See below for detail.

File: package.json (hunk 5 above)
Comment (ECHO-HELLO-WORLD424) [reply package.json]: Now all MCP features are turned off by default, that should be safer. see below for details.

File: src/mcp/translation.ts (hunk 4 above)
Comment (ECHO-HELLO-WORLD424) [reply src/mcp/translation.ts]: Fixed. Changelogs below.

File: package.json [hunk 9]
```diff
@@ -1039,6 +1101,8 @@
 		"typescript-eslint": "^8.26.0"
 	},
 	"dependencies": {
+		"@cfworker/json-schema": "4.1.1",
```
[the remark this answers is not in the document]
Comment (ECHO-HELLO-WORLD424) [reply package.json]: Moved input validation to `rce.ts`. Details below

File: package.json (hunk 3 above)
Comment (ECHO-HELLO-WORLD424) [reply package.json]: Fixed.

File: CHANGELOG.md [hunk 10]
```diff
@@ -18,6 +18,19 @@ Hello Neuro! If you're reading this, it means Vedal has let you read the changel
...
+  - Currently supports single MCP server connection via HTTP transport, on desktop (works) and web (buggy as browser block cross-origin requests)
+  - MCP tools are controlled via the `neuropilot.permission.mcpTools` permission (Copilot/Autopilot/Off(default)).
+  - New settings: `neuropilot.mcp.serverUrl`, `neuropilot.mcp.timeout`, `neuropilot.mcp.autoConnect`.
+  - New commands: Connect to MCP Server, Disconnect from MCP Server, Refresh MCP Tools, Show MCP Status, Test MCP Tool Call (Debug).
+  - New context sent to Neuro (to notify her about MCP tools. The src/mcp/translation.ts will add `mcp_` prefix to translated MCP tools )
+    - On MCP connected: "An MCP server with xx tools is now connected. You may access these tools whose names start with `mcp_`."
+    - On MCP disconnected: "The MCP server is now disconnected. You can't access the tools whose names start with `mcp_` anymore."
+  - Known limitations: 
+    - No multi-server support 
+    - `/mcp` endpoint only
```
Comment (KTrain5169) [inline CHANGELOG.md]: What does this mean?

File: src/web/extension.ts (hunk 6 above)
Comment (ECHO-HELLO-WORLD424) [reply src/web/extension.ts]: Fixed this

File: src/web/extension.ts (hunk 7 above)
Comment (ECHO-HELLO-WORLD424) [reply src/web/extension.ts]: Also fixed this

File: CHANGELOG.md [hunk 11]
```diff
@@ -18,6 +18,19 @@ Hello Neuro! If you're reading this, it means Vedal has let you read the changel
...
+  - MCP tools are controlled via the `neuropilot.permission.mcpTools` permission (Copilot/Autopilot/Off(default)).
+  - New settings: `neuropilot.mcp.serverUrl`, `neuropilot.mcp.timeout`, `neuropilot.mcp.autoConnect`.
+  - New commands: Connect to MCP Server, Disconnect from MCP Server, Refresh MCP Tools, Show MCP Status, Test MCP Tool Call (Debug).
+  - New context sent to Neuro (to notify her about MCP tools. The src/mcp/translation.ts will add `mcp_` prefix to translated MCP tools )
+    - On MCP connected: "An MCP server with xx tools is now connected. You may access these tools whose names start with `mcp_`."
+    - On MCP disconnected: "The MCP server is now disconnected. You can't access the tools whose names start with `mcp_` anymore."
+  - Known limitations: 
+    - No multi-server support 
+    - `/mcp` endpoint only 
+    - No individual tool-level control.
```
[the remark this answers is not in the document]
Comment (ECHO-HELLO-WORLD424) [reply CHANGELOG.md]: You mean the limitations part?

File: CHANGELOG.md [hunk 12]
```diff
@@ -27,11 +27,11 @@ Hello Neuro! If you're reading this, it means Vedal has let you read the changel
...
     - On MCP disconnected: "The MCP server is now disconnected. You can't access the tools whose names start with `mcp_` anymore."
   - MCP actions will automatically be registered or unregistered if the server is connected/disconnected.
-  - Known limitations: 
-    - No multi-server support 
-    - `/mcp` endpoint only 
-    - No individual tool-level control.
-    - **Web version requires MCP servers with CORS headers** (browsers block cross-origin requests without proper `Access-Control-Allow-Origin` headers).
+  - Known limitations:
+    - No multi-server support (only one MCP server can be connected at a time)
+    - Only supports Streamable-HTTP transport with `/mcp` endpoint (no `stdio` or legacy `/sse` SSE transport)
```
Comment (KTrain5169) [inline CHANGELOG.md]: Is this (supporting only `/mcp` subpath) a limitation of the MCP SDK? Pretty weird limitation if so...

File: CHANGELOG.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (KTrain5169) [reply CHANGELOG.md]: Individual tool-level control - as in you could disable/enable them similar to `actions.disabledActions`.

It's alright if you don't want to though

[the remark this answers is not in the document]
Comment (ECHO-HELLO-WORLD424) [reply CHANGELOG.md]: Oh forgot to reply this. I removed it

[the remark this answers is not in the document]
Comment (ECHO-HELLO-WORLD424) [reply CHANGELOG.md]: Implemented. See below for details
